apptest/testcase.go

```diff
@@ -103,6 +104,50 @@ func (tc *TestCase) MustStartVminsert(instance string, flags []string) *Vminsert
 	return app
 }
 
+type vmcluster struct {
+	*Vminsert
+	*Vmselect
+	vmstorages []*Vmstorage
+}
+
+func (c *vmcluster) ForceFlush(t *testing.T) {
+	time.Sleep(2 * time.Second)
```

Comment:
Done. So I introduced 2 mechanisms to address test flakiness:

1.  Waiting in vminsert. vminsert does not push records to vmstorage right away. Instead, it buffers the accepted records for approximately [200ms](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0e9767cd77a601a2478079a6e8ad4c0327aed790/app/vminsert/netstorage/netstorage.go#L159). When it finally sends the data, it [increments](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0e9767cd77a601a2478079a6e8ad4c0327aed790/app/vminsert/netstorage/netstorage.go#L308) the corresponding `vm_rpc_rows_sent_total` metric. This can be used for waiting. The implementation reads the metric several times until it is increased by at least the expected value. Note that such implementation becomes unreliable if the data is inserted concurrently (which is not our case at the moment, but this needs to be taken into account when the tests are written such that write data into vminsert from multiple goroutines).

2. The TestCase has been extended with the Assert() method that compares the values with retry. This is close what you proposed the other day: retrieve the metric count until it matches the expected value (or fail on timeout). The difference is that it allows to wait for any result. Because there can be some tests when you just can't count metrics because, for example, there is no a vmselect instance.

With this Assert() method the waiting in vminsert may not be needed, especially that this waiting may be unreliable. However, if we don't wait, then storage.DebugFlush() becomes useless (because there is nothing to flush). And without DebugFlush() the tests will spend way more time waiting for the data to become available. So this waiting in vminsert and the subsequent DebugFlush() in vmstorage is an optimization.

Further optimization can be done by making the metric update period configurable in httpserver. Currently it is a hardcoded value of [1 second](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/0e9767cd77a601a2478079a6e8ad4c0327aed790/lib/appmetrics/appmetrics.go#L39). Which causes waiting in vminsert to be much longer than the push delay of 200ms. We could make this time configurable via a flag, say `--metrics.cacheDuration`.